pkg/log/context.go

```diff
@@ -23,92 +23,145 @@ import (
+	hook := filename.NewHook()
+	hook.Field = FieldComponentName
+	defaultEntry.Logger.AddHook(hook)
+	defaultEntry = defaultEntry.WithField(FieldCorrelationID, "-")
+	_, err := Configure(context.Background(), currentSettings)
+	if err != nil {
+		panic(err)
+	}
+}
 
 // Settings type to be loaded from the environment
 type Settings struct {
-	Level  string    `description:"minimum level for log messages"`
-	Format string    `description:"format of log messages. Allowed values - text, json"`
-	Output io.Writer `mapstructure:"-"`
+	Level  string `description:"minimum level for log messages" json:"level,omitempty"`
+	Format string `description:"format of log messages. Allowed values - text, json" json:"format,omitempty"`
+	Output string `description:"output for the logs. Allowed values - /dev/stdout, /dev/stderr, ginkgowriter" json:"output,omitempty"`
 }
 
 // DefaultSettings returns default values for Log settings
 func DefaultSettings() *Settings {
 	return &Settings{
 		Level:  "error",
 		Format: "text",
-		Output: os.Stdout,
+		Output: os.Stdout.Name(),
 	}
 }
 
 // Validate validates the logging settings
 func (s *Settings) Validate() error {
 	if _, err := logrus.ParseLevel(s.Level); err != nil {
-		return fmt.Errorf("validate Settings: %s", err)
+		return fmt.Errorf("validate Settings: log level %s is invalid: %s", s.Level, err)
 	}
+
 	if len(s.Format) == 0 {
-		return fmt.Errorf("validate Settings: LogFormat missing")
+		return fmt.Errorf("validate Settings: log format missing")
 	}
-	if s.Output == nil {
-		return fmt.Errorf("validate Settings: LogOutput missing")
+
+	if _, ok := supportedFormatters[s.Format]; !ok {
+		return fmt.Errorf("validate Settings: log format %s is invalid", s.Format)
+	}
+
+	if len(s.Output) == 0 {
+		return fmt.Errorf("validate Settings: log output missing")
 	}
+
+	if _, ok := supportedOutputs[s.Output]; !ok {
+		return fmt.Errorf("validate Settings: log output %s is invalid", s.Output)
+	}
+
 	return nil
 }
 
 // Configure creates a new context with a logger using the provided settings.
-func Configure(ctx context.Context, settings *Settings) context.Context {
-	once.Do(func() {
-		level, err := logrus.ParseLevel(settings.Level)
-		if err != nil {
-			panic(fmt.Sprintf("Could not parse log level configuration: %s", err))
-		}
-		formatter, ok := supportedFormatters[settings.Format]
-		if !ok {
-			panic(fmt.Sprintf("Invalid log format: %s", settings.Format))
-		}
-		logger := &logrus.Logger{
-			Formatter: formatter,
-			Level:     level,
-			Out:       settings.Output,
-			Hooks:     make(logrus.LevelHooks),
-		}
-		hook := filename.NewHook()
-		hook.Field = FieldComponentName
-		logger.AddHook(hook)
-		defaultEntry = logrus.NewEntry(logger)
-		defaultEntry = defaultEntry.WithField(FieldCorrelationID, "-")
-	})
-	return ContextWithLogger(ctx, defaultEntry)
+func Configure(ctx context.Context, settings *Settings) (context.Context, error) {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	currentSettings = settings
```

Comment:
if the provided settings are invalid, then until they're reconfigured you'll be receiving invalid settings on getConfig. This is only valid when modifying the file, but still.

Reply:
we could add another validation of the settings 
`	if err := settings.Validate(); err != nil {
		return nil, err
	}
`

at the start of log.Configure